docs/tech-notes/index-backfill.md

```diff
@@ -199,3 +199,186 @@ tombstone when it expects a `9` and fails. When retried transactionally, the
+
+Currently the schema change advances an index to the DELETE_AND_WRITE state
+before backfilling -- this means that CRUD ops will write index entries, and
+these writes and the backfill writes are both transactional and thus protected
+against the usual anomalies. If a query were to try to write a duplicate value,
+one of either the index backfill write or the query write will succeed and the
+other will (perhaps after being pushed/retried), get a conflict from its InitPut
+and yield a uniqueness violation error.
+
+AddSSTable however adds its batch of keys, which can each have any timestamp,
+without consulting the timestamp cache, potentially rewriting history out from
+under prior reads or shadowing existing keys at lower timestamps. This means the
+backfiller could add a batch of entries including clobbering an entry written by
+a query when one of the two should have instead failed.
+
+### Rejected Fix: Teach AddSSTable to be transactional
+
+We've looked at, and opted against, trying to teach AddSSTable to be more
+transactional -- the point of it is to be cheaper than KV, and teaching it KV's
+fancy tricks would likely start to defeat that purpose. For example, if we e.g.
+teach it to check the timestamp cache (and optionally ensure the keys in the SST
+match the command time), retrying a whole SST because of a conflicting is
+expensive, and again, our goal is to make this cheaper.
+
+### Rejected Fix: Teach AddSSTable to behave like InitPut w.r.t existing keys
+
+Furthermore, we'd need more than just proper transactional semantics -- the
+current index backfill relies on the InitPut semantics to correctly return
+uniqueness conflict errors. Here again, we considered, and rejected, teaching
+AddSST to support those semantics. Doing so would likely use a flag that added a
+check the current iteration of the underlying range (for computing merged MVCC
+stats) to check if any of the existing keys conflicted with keys in the SST --
+and if so, return an error instead of ingesting it. However, as with
+transactional isolation, it isn't clear that we actually want to reject the
+whole SST. A rejected SST would be retried -- it might then find an actual
+anomaly, or it might find that a query's point-write changed a stored column or
+something. Retrying might have other challenges -- a) we want to make big SSTs
+to minimize overhead, but that maximizes retry cost and also maximizes the
+chance a given SST will require a retry -- with it eventually approaching a
+certainty. It also isn't clear how we'd actually do the retry -- e.g. in a
+dist-sort based approach, the chunk readers are divorced from the actual SST
+that was rejected, so we'd need to figure that out. And, at a high level, the
+simplicity of ingestion is what makes AddSSTable fast. The more we mess with it,
+the more likely it could lose the attribute that makes it appealing in the first
+place.
+
+### Conflict and Anomaly Fix: Counting Entries
```

Comment:
An idea I've been kicking around is abusing the logical part of the hlc timestamp for this. If you could somehow guarantee that every sst in a backfill gets a unique int32 and construct each sst with that int32 as the logical part of the hlc time (presumably the walltime is the same across everything in the backfill). We already iterate both the sst and existing data in AddSSTable to construct the stats, you could detect conflicts at the same time by looking for any keys that already exist but with a different timestamp (they could exist with the same timestamp if the command got retried).

Reply:
We'd discussed a couple ways we could use the iterate during addsst to look for matching keys, but not all matching keys are actual uniqueness violations though -- e.g. a SQL write to a row changing the value of a column stored in that index would write an entry for the same key at a higher timestamp with a different value.

If we turn the matching key into an error at addsst, similar to existing backfill, the caller needs to retry the whole chunk to see if the conflict goes away? Given the goal of ssd-ingestion is to build and ingest big chunks, a table with enough traffic seems likely to end up never making progress in this situation unless I'm missing something?

That said, just in terms of wasted work, I'm not thrilled about doing the whole backfill before we realize the very first key was a dupe. We could do the iterate-time dupe detection but we don't have to make dupes into errors: maybe we just look for them, record them and send them back with the success message so the backfiller can then put them on a queue of suspect keys to check. A parallel a process would pull suspect keys to check if they are actual uniqueness violations vs properly layered CRUD writes and could abort the job more promptly in the former case to avoid so much wasted work.